File: typescript/2023/src/day5/day5.ts
import { readAsString } from "../utils/getInput";

const getSettings = (input: string): [number[], number[][][]] => {
  const [seedsString, ...maps] = input.split("\n\n");
  const seeds = seedsString
    .split(":")[1]
    .trim()
    .split(" ")
    .map((val) => parseInt(val));
  const settings = maps
    .map((mapSection) => mapSection.split("\n").slice(1))
    .map((mapSection) =>
      mapSection.map((line) => line.split(" ").map((val) => parseInt(val))),
    );
  return [seeds, settings];
};

const runMap = (settings: number[][], seeds: number[]) => {
  for (let index in seeds) {
    const seed = seeds[index];
    settings.forEach((line) => {
      const [destStart, sourceStart, range] = line;
      if (seed >= sourceStart && seed < sourceStart + range) {
        seeds[index] = destStart + (seed - sourceStart);
      }
    });
  }
};

const runMapRange = (settings: number[][], seeds: number[][]): number[][] => {
  const mappedSeeds: number[][] = [];
  let unmappedSeeds: number[][] = [];
  settings.forEach((line) => {
    unmappedSeeds = [];
    const [destStart, sourceStart, range] = line;
    const sourceEnd = sourceStart + range;
    seeds.forEach((seed) => {
      const seedStart = seed[0];
      const seedEnd = seed[1];
      const start = [seedStart, Math.min(seedEnd, sourceStart)];
      const mid = [
        Math.max(seedStart, sourceStart),
        Math.min(seedEnd, sourceEnd),
      ];
      const end = [Math.max(sourceEnd, seedStart), seedEnd];
      if (start[1] > start[0]) unmappedSeeds.push(start);
      if (mid[1] > mid[0])
        mappedSeeds.push([
          mid[0] - sourceStart + destStart,
          mid[1] - sourceStart + destStart,
        ]);
      if (end[1] > end[0]) unmappedSeeds.push(end);
    });
    seeds = unmappedSeeds;
  });

  return mappedSeeds.concat(seeds);
};

const expandSeeds = (seeds: number[]): number[][] => {
  const expanded = [];
  for (let i = 0; i < seeds.length; i += 2) {
    expanded.push([seeds[i], seeds[i] + seeds[i + 1] - 1]);
  }
  return expanded;
};

export const partOne = (filename: string): number => {
  const input = readAsString(filename);
  let [seeds, settings] = getSettings(input);
  settings.forEach((setting) => runMap(setting, seeds));
  // WHY DOES SORT CONVERT TO STRINGS??????????????
  return seeds.sort((a, b) => a - b)[0];
};

export const partTwo = (filename: string): number => {
  const input = readAsString(filename);
  const [seeds, settings] = getSettings(input);
  let seedRanges = expandSeeds(seeds);
  settings.forEach(
    (setting) => (seedRanges = runMapRange(setting, seedRanges)),
  );
  // WHY DOES SORT CONVERT TO STRINGS??????????????
  return seedRanges.map((range) => range[0]).sort((a, b) => a - b)[0];
};

File: golang/2023/day5/day5.go
package day5

import (
	"aoc/golang/utils"
	"fmt"
	"strconv"
	"strings"
	"time"
)

type AlmanacMap []Conversions
type Seeds []int
type Conversions []int
type SeedRange []int

func getSettings(input string) (Seeds, []AlmanacMap) {
	settings := strings.Split(input, "\n\n")
	seedsStr, settings := strings.Fields(strings.Split(settings[0], ":")[1]), settings[1:]

	seeds := make(Seeds, 0, len(seedsStr))
	for _, s := range seedsStr {
		v, _ := strconv.Atoi(s)
		seeds = append(seeds, v)
	}

	var maps []AlmanacMap
	for _, mapSection := range settings {
		var section []Conversions
		mapSection := strings.TrimSpace(mapSection) // Fix line feeds
		for _, line := range strings.Split(mapSection, "\n")[1:] {
			var sectionVals []int
			for _, val := range strings.Fields(line) {
				intVal, _ := strconv.Atoi(val)
				sectionVals = append(sectionVals, intVal)
			}
			section = append(section, sectionVals)
		}

		maps = append(maps, section)
	}

	return seeds, maps
}

func runMap(seeds *Seeds, maps AlmanacMap) {
	for i, s := range *seeds {
		for _, m := range maps {
			dest, source, rnge := m[0], m[1], m[2]
			if s >= source && s < source+rnge {
				(*seeds)[i] = (s - source) + dest

			}
		}
	}
}

func runMapRanges(seeds []SeedRange, maps AlmanacMap) []SeedRange {
	var unmappedSeeds []SeedRange
	var mappedSeeds []SeedRange

	for _, m := range maps {
		unmappedSeeds = []SeedRange{}
		dest, source, rnge := m[0], m[1], m[2]
		sourceEnd := source + rnge

		for _, s := range seeds {
			seedStart, seedEnd := s[0], s[1]
			start := SeedRange{seedStart, min(source, seedEnd)}
			mid := SeedRange{max(seedStart, source), min(seedEnd, sourceEnd)}
			end := SeedRange{max(sourceEnd, seedStart), seedEnd}
			if start[1] > start[0] {
				unmappedSeeds = append(unmappedSeeds, start)
			}
			if mid[1] > mid[0] {
				mappedSeeds = append(mappedSeeds, SeedRange{(mid[0] - source) + dest, (mid[1] - source) + dest})
			}
			if end[1] > end[0] {
				unmappedSeeds = append(unmappedSeeds, end)
			}
		}
		seeds = unmappedSeeds
	}
	return append(mappedSeeds, seeds...)
}

func expandSeeds(seeds Seeds) []SeedRange {
	var expandedSeeds []SeedRange

	for i := 0; i < len(seeds); i += 2 {
		expandedSeeds = append(expandedSeeds, []int{seeds[i], seeds[i] + seeds[i+1] - 1})
	}
	return expandedSeeds
}

func minSeed(seeds Seeds) int {
	minSeed := seeds[0]
	for _, s := range seeds {
		if s < minSeed {
			minSeed = s
		}
	}
	return minSeed
}

func minSeedRange(seeds []SeedRange) int {
	minSeed := seeds[0][0]

	for _, s := range seeds {
		if s[0] < minSeed {
			minSeed = s[0]
		}
	}
	return minSeed
}

func runPartOne(filename string) int {
	fmt.Println("Part One")
	input := utils.ReadAsString(filename)
	seeds, maps := getSettings(input)
	for _, m := range maps {
		runMap(&seeds, m)
	}
	return minSeed(seeds)
}

func runPartTwo(filename string) int {
	start := time.Now()
	fmt.Println("Part Two")
	input := utils.ReadAsString(filename)
	seeds, maps := getSettings(input)
	seedsRanges := expandSeeds(seeds)
	for _, m := range maps {
		seedsRanges = runMapRanges(seedsRanges, m)
	}
	fmt.Println(time.Since(start))
	return minSeedRange(seedsRanges)
}
